File: 37数字在排序数组中出现的次数/Solution.cpp

class Solution {
public:
    int searchfirst(int num, vector<int>& vec){
        int begin = 0;
        int end = vec.size()-1;
        int middle;
        while(begin < end - 1){
            middle = begin+(end-begin)/2;
            if(vec[middle] >= num){
                end = middle;
            }else
                begin = middle;
        }
        if(vec[begin] == num) return begin;
        else return end;
 }
 
 int searchback(int num, vector<int>& vec){
        int begin = 0;
        int end = vec.size()-1;
        int middle;
        while(begin < end - 1){
            middle = begin+(end-begin)/2;
            if(vec[middle] > num){
                end = middle;
            }else
                begin = middle;
        }
        if(vec[end] == num) return end;
        else return begin;

 }
  int GetNumberOfK(vector<int> data ,int k) {
        int length = data.size();
        if(length == 0){
            return 0;
        }
        int first = searchfirst(k,data);
        int last = searchback(k, data);
        if(first != -1 && last != -1){
            return last - first + 1;
        }
        return 0;
    }
}
/* 这道题时间复杂度最优的方法一定是二分，这就涉及到二分的一个进阶
就是寻找重复目标数的第一个和最后一个位置，区别在于第一个则在找
middle前找到时，不跳出，继续找，最后一个是在middle后找到时，不跳出
继续找 */

File: 37数字在排序数组中出现的次数/Solution.java
public class Solution {
    public int GetNumberOfK(int [] nums , int K) {
       int first = binarySearch(nums, K);
    int last = binarySearch(nums, K + 1);
    return (first == nums.length || nums[first] != K) ? 0 : last - first;
}

private int binarySearch(int[] nums, int K) {
    int l = 0, h = nums.length;
    while (l < h) {
        int m = l + (h - l) / 2;
        if (nums[m] >= K)
            h = m;
        else
            l = m + 1;
    }
    return l;
    }
}
/* 这个解法更妙，只用找first，然后直接找k+1的位置即可
哪怕k+1不存在，也可以返回重复目标数后面的第一个数
二分法牛逼！ */